design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,219 @@
+
+Cons:
+* We’re adding yet another semi alpha knob to an already complicated set of knobs that each VM owner needs to be aware of.
+
+## Alternative #2 Configure both pod and Guest
+It is possible to use both the current Guest API and the old yet existing pod API, like so:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        resources:
+          requests:
+            cpu: '10'
+          limits:
+            cpu: '10'
+        cpu:
+          cores: 4
+          sockets: 1
+          threads: 2
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+```
+In order to do that the current validating webhooks regarding dedicatedCpuPlacement and IsolatedEmulatorThread will have to be loosened.
+
+Pros:
+* This option does not introduce any alpha knobs.
+
+Cons:
+* This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse. Loosening the validating webhook doesn't make sense in the general case.
+* Implementation difficulty: KubeVirt will need to adjust internal assumptions that are based on the fact that VMs {req, limits}.cpu = vCPUs. Specifically, how the housekeeping cgroup is being formed, but possibly other aspects.
+
+## Scalability
+The solution proposed is “wasting” one unused CPU for the purpose of maintaining SMT alignment.
+
+## Update/Rollback Compatibility
+Since the annotation is not an API change, it is backportable.
+When a better solution is possible, Kubevirt can simply replace the annotation support with a better approach, thus not making the change relatively seamless for the client. 
+
+## Functional Testing Approach
+Will add unit tests to `renderresources_test.go` to check extra housekeeping logic, and in `templates.go` to check the virt-launcher pod spec output.
+
+# Implementation Phases
+## Phase 1: Align CPU to SMT requirements when feature gate is enabled and annotation is set on the VMI.
+* Virt-config:
+  * Add support for the new feature gate: `EmulatorThreadCompleteToEvenParity`.
+* Virt-API:
+  * Add support for the new alpha annotation `kubevirt.io/EmulatorThreadCompleteToEvenParity:`.
+  * in case annotation is set on kubevirt CR, the feature gate is enabled and the isolateEmulatorThread is enabled:
+    * the VMI mutator webhook will copy the kubevirt CR annotation to the VMI annotations.
+  * in case the VMI has the annotation and either the feature gate is disabled or the isolateEmulatorThread is disabled:
+    * the VMI creation will be blocked. 
+* Virt-controller:
+  * in case isolateEmulatorThread is enabled and the annotation is set on the VMI:
+    * Add logic to Align the total number of CPUs to be even.
+## Phase 2: Adjust housekeeping cgroup to use the extra CPU added by the alignment, if such a one is available.
```

Comment:
Why is phase 2 important? Is single CPU typically not more than enough for housekeeping?

Reply:
Thanks for the pointer. I would add a sentence about the reason in the text of the proposal. My opinion - phase 2 does not worth our trouble.

Reply:
I understand your view, but note that this part is mostly already merged: 
- https://github.com/kubevirt/kubevirt/pull/10783 merged
- https://github.com/kubevirt/kubevirt/pull/10839 small fix lgtmed, waiting for approval

do you think it is worth reverting?

Reply:
No. I don't see this as particularly harmful. Just more code for us to maintain, with negligible benefit.

Reply:
As my suggestion to keep this optional/not-a-must-for-now was rejected, it seems it is not the opinion of our maintainers.

Reply:
Please add a text here explaining why it was important

Reply:
At the time @vladikr thought that not incorporating  the second CPU for cgroups is a waist of CPU power.